docs/design/byos.md

```diff
@@ -1,13 +1,153 @@
+Later, we may consider whether this same solution works for:
+- AKS
+- Any Kubernetes
+- Other clouds
+
+Under the hood, we need an image, an ARM template, an Azure subscription, and instructions about how much capacity to provision.
+That doesn't require us to expose the full complexity of ARM templates and Azure subscriptions to every customer.
+
+### Setup
+
+For a lot of customers, it would be enough to have them
+- go to pool setup
+- create a new pool of type "Custom Azure-hosted" (name t.b.d.)
+- pick one of a few different Azure VM SKUs + a few different agent lifetime policies
+- pick their Azure subscription
+- have Azure Pipelines configure it all for them
+
+For the advanced customer who really needs all the customization available, they'll switch to actually giving us the ARM template and so on.
+
+_TODO_: draw pictures
+
+### Infrastructure problems
+
+_TODO_: How will we let the adminstrator know of any problems that occur?
+Azure Portal will report things, but the leading indicator is probably "my builds aren't running".
+
+## Alternatives considered
+
+### Custom / configurable SKU
+
+Offering additional Azure SKUs in Hosted will meet some of the above scenarios.
+Specifically, if all you need are additional resources (memory, compute, etc.), then a custom SKU works.
+
+Pros:
+- Easiest for the customer to understand
+- Potential business model around premium SKUs
+
+Cons:
+- Capacity planning and buildout becomes a lot more complex
+- If a customer wants a SKU we don't offer, they're out of luck until we add it
+- Does not offer custom elasticity, network environment, or image
+
+### Bring your own cloud (pool providers)
+
+Putting the entire agent alloc/release cycle in the hands of the customer offers them ultimate flexibility.
+
+Pros:
+- Maximum flexibility
+- Customer can select other infrastructure: on-prem, in a different hosting provider, etc.
+
+Cons:
+- Much heavier burden to understand, set up, and maintain
+- Customer has to swallow full complexity load to get any flexibility
+
+## "Clickstops" along the way
+
+Of course, we won't want to do a huge multi-month effort and then dump it on the community at once.
+We can deliver several intermediate levels of value on the way to a full solution.
+At every point, we can re-evaluate if the decisions we're making match customer needs at the time.
+
+_TODO_: add a new set of intermediate deliverables
+
+<!-- this generated more controversy than agreement, so needs to be re-thought
+**Stage 0**. Using a customer's Azure subscription, run pipelines against our native images using Hosted-like "throwaway" VMs.
+The customer tells us exactly how many agents to keep around.
+The benefits in this stage primarily go to us: we gain experience running a multi-tenant service against subscriptions we don't own.
+This would be a private preview rather than a generally available feature.
+
+**Stage 1**. Offer the ability to keep a Hosted-like agent alive for multiple consecutive runs.
```

Comment:
Could https://azure.microsoft.com/en-us/blog/ephemeral-os-disk-limited-public-preview/ be of any help ?
I would not mind having "x" agents always on, and if any new pipeline requests an agent beyond x while all x are running a job, then the infrastructure would provision new agents till a max limit.
We only care about having the machine in state we created every time we run a job.

Reply:
We cut all the discussion of "intermediate steps" for now. That doesn't mean we'll land the feature without some preview programs, just that this wasn't the right list of steps.